toolchain/check/impl_validation.cpp

```diff
@@ -244,90 +239,104 @@ static auto DiagnoseFinalImplsOverlapOutsideMatchFirst(Context& context,
   // than quadratic time. From @zygoloid: Possibly by converting the set of
   // impls into a decision tree.
-  for (auto [index_a, impl_a_id] : llvm::enumerate(impl_ids)) {
-    auto impl_a = GetImplInfo(context, impl_a_id);
-
-    if (impl_a.is_final && impl_a.is_local) {
-      // =======================================================================
-      /// Rules for an individual final impl and its internal values.
-      // =======================================================================
-      if (DiagnoseFinalImplNotInSameFileAsRootSelfTypeOrInterface(
-              context, impl_a, interface_ir_id)) {
+  //
+  // For each impl, we compare it pair-wise which each impl found before it, so
+  // that diagnostics are attached to the later impl, as the earlier impl on its
+  // own does not generate a diagnostic.
+  size_t num_impl_ids = impls_and_interface.size();
+  for (auto [split_point, impl_b_id] :
+       llvm::drop_begin(llvm::enumerate(impl_ids))) {
+    auto impl_b = GetImplInfo(context, impl_b_id);
+
+    // Prevent diagnosing the same error multiple times for the same `impl_b`
+    // against different impls before it. But still ensure we do give one of
+    // each diagnostic when they are different errors.
+    bool did_diagnose_non_final_impls_with_same_type_structure = false;
+    bool did_diagnose_unmatchable_non_final_impl_with_final_impl = false;
+    bool did_diagnose_final_impls_overlap_in_different_files = false;
+    bool did_diagnose_final_impls_overlap_outside_match_first = false;
+
+    auto impls_before = llvm::drop_end(impl_ids, num_impl_ids - split_point);
+    for (auto impl_a_id : impls_before) {
+      auto impl_a = GetImplInfo(context, impl_a_id);
+
+      // Only enforce rules when at least one of the impls was written in this
+      // file.
+      if (!impl_a.is_local && !impl_b.is_local) {
         continue;
       }
-    }
-
-    for (auto impl_b_id : llvm::drop_begin(impl_ids, index_a + 1)) {
-      auto impl_b = GetImplInfo(context, impl_b_id);
 
       if (!impl_a.is_final && !impl_b.is_final) {
         // =====================================================================
         // Rules between two non-final impls.
-        //
-        // It's possible to write invalid non-final impls in two files that do
-        // not know about each other, and they are only invalid when they are
-        // both imported and visible together. So we need to compare non-local
-        // impls with each other here.
         // =====================================================================
-        if (DiagnoseNonFinalImplsWithSameTypeStructure(context, impl_a,
-                                                       impl_b)) {
-          continue;
+        if (!did_diagnose_non_final_impls_with_same_type_structure) {
+          if (DiagnoseNonFinalImplsWithSameTypeStructure(context, impl_a,
+                                                         impl_b)) {
```

Comment:
```suggestion
        if (!did_diagnose_non_final_impls_with_same_type_structure &&
            impl_a.is_local && impl_b.is_local &&
            DiagnoseNonFinalImplsWithSameTypeStructure(context, impl_a,
                                                       impl_b)) {
```

Reply:
You're saying here that two impls can have the exact same type structure if they are in different files? Then how are they prioritized?

And doesn't that again break file concatenation, where code works before concat but then fails afterward?

Reply:
Sorry, I meant to say they **can't** have the exact same type structure if they are in different files.

Reply:
Oh I guess this is saying the orphan rule prevents overlap of this nature across files so we don't need to diagnose that:

> Orphan rule: Some name from the type structure of an impl declaration must be defined in the same library as the impl, that is some name must be local.

However.. then we do break file concatenation.. :/

Reply:
Or, maybe not. I guess the orphan rule means when we concat they will have different type structures. So this change just avoiding looking for things that are impossible (once orphan rule is implemented) and thus concat will work. Okay, cool.